### PR TITLE
Raven db 8835

### PR DESCRIPTION
**Pull request by maximburyak**

### CODE REVIEW

File: src/Raven.Server/Documents/Queries/AST/JavascriptCodeQueryVisitor.cs [hunk 1]
```diff
@@ -95,9 +96,92 @@ public override void VisitCompoundWhereExpression(BinaryExpression @where)
 
         public override void VisitMethod(MethodExpression expr)
         {
+            // The regex usage here is just temporary, to get things working. Ideally, we want to implement the string functions in Jint
+            if (expr.Name.Value.Equals("startswith", StringComparison.OrdinalIgnoreCase) && expr.Arguments.Count == 2)
+            {
+                _sb.Append("new Regexp('^");
```
Comment (ayende) [inline src/Raven.Server/Documents/Queries/AST/JavascriptCodeQueryVisitor.cs:102]: Javascript _has_ startsWith method

File: src/Raven.Server/Documents/Queries/AST/JavascriptCodeQueryVisitor.cs [hunk 2]
```diff
@@ -95,9 +96,92 @@ public override void VisitCompoundWhereExpression(BinaryExpression @where)
...
+                _sb.Append("').test('");
+                VisitExpression(expr.Arguments[0]);
+                
+                _sb.Append("')");
+                return;
+            }
+            
+            if (expr.Name.Value.Equals("endswith", StringComparison.OrdinalIgnoreCase) && expr.Arguments.Count == 2)
+            {
+                _sb.Append("new Regexp('");
```
Comment (ayende) [inline src/Raven.Server/Documents/Queries/AST/JavascriptCodeQueryVisitor.cs:114]: Same here for endsWith

Comment (ayende) [reply src/Raven.Server/Documents/Queries/AST/JavascriptCodeQueryVisitor.cs:114]: However, we need to be sure that these are case insensitive, so we will introduce them as methods in `ScriptRun`

File: src/Raven.Server/Documents/Queries/AST/JavascriptCodeQueryVisitor.cs [hunk 3]
```diff
@@ -95,9 +96,92 @@ public override void VisitCompoundWhereExpression(BinaryExpression @where)
...
+                var param = expr.Arguments[1] as ValueExpression;
+                _sb.Append(Regex.Escape(param.Token.Value));
+                _sb.Append("$').test('");
+                VisitExpression(expr.Arguments[0]);
+                
+                _sb.Append("')");
+                return;
+            }
+            
+            if (expr.Name.Value.Equals("intersect", StringComparison.OrdinalIgnoreCase) && expr.Arguments.Count >= 2)
```
Comment (ayende) [inline src/Raven.Server/Documents/Queries/AST/JavascriptCodeQueryVisitor.cs:124]: I don't understand how that is supposed to work

File: src/Raven.Server/Documents/Queries/AST/JavascriptCodeQueryVisitor.cs [hunk 4]
```diff
@@ -95,9 +96,92 @@ public override void VisitCompoundWhereExpression(BinaryExpression @where)
...
+                        _sb.Append(" && ");
+                }
+                
+                _sb.Append("')");
+                return;    
+            }
+            
+            if (expr.Name.Value.Equals("regex", StringComparison.OrdinalIgnoreCase) && expr.Arguments.Count == 2)
+            {
+                _sb.Append("new Regexp(' ");
```
Comment (ayende) [inline src/Raven.Server/Documents/Queries/AST/JavascriptCodeQueryVisitor.cs:143]: Better to implement it ourselves, we need to make sure that it has the same beahvior as the where clause.

File: src/Raven.Server/Documents/Queries/AST/JavascriptCodeQueryVisitor.cs [hunk 5]
```diff
@@ -95,9 +96,92 @@ public override void VisitCompoundWhereExpression(BinaryExpression @where)
...
+            if (expr.Name.Value.Equals("exists", StringComparison.OrdinalIgnoreCase) && expr.Arguments.Count == 1)
+            {
+                _sb.Append("(typeof "); 
+                VisitExpression(expr.Arguments[0]);
+                _sb.Append("=== 'undefined')");
+                return;    
+            }
+            
+            if (expr.Name.Value.Equals("exact", StringComparison.OrdinalIgnoreCase) && expr.Arguments.Count ==2)
+            {
```
Comment (ayende) [inline src/Raven.Server/Documents/Queries/AST/JavascriptCodeQueryVisitor.cs:160]: It is better to not support this.

File: src/Raven.Server/Documents/Queries/AST/JavascriptCodeQueryVisitor.cs [hunk 6]
```diff
@@ -95,9 +96,92 @@ public override void VisitCompoundWhereExpression(BinaryExpression @where)
...
+                VisitExpression(expr.Arguments[0]);
+                _sb.Append(" == ");
+                VisitExpression(expr.Arguments[1]);
+                return;
+            }
+            
+            if (expr.Name.Value.Equals("search", StringComparison.OrdinalIgnoreCase) && expr.Arguments.Count == 2)
+            {
+                _sb.Append("new Regexp('");
```
Comment (ayende) [inline src/Raven.Server/Documents/Queries/AST/JavascriptCodeQueryVisitor.cs:169]: It is better to not support this and error

File: test/FastTests/Client/Subscriptions/RavenDB-8404.cs [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+                    amre.Set();
+                });
+                
+                Assert.True(await amre.WaitAsync(_reasonableWaitTime));
+                Assert.Equal(3, users.Count);
+            }
+        }
+        
+        [Fact]
+        public async Task SubscriptionsRQLSupportStartsWith2()
```
Comment (ayende) [inline test/FastTests/Client/Subscriptions/RavenDB-8404.cs:59]: bad test name

File: test/FastTests/Client/Subscriptions/RavenDB-8404.cs [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+                    users.AddRange(x.Items.Select(i => i.Result));
+                    amre.Set();
+                });
+                
+                Assert.True(await amre.WaitAsync(_reasonableWaitTime));
+                Assert.Equal(3, users.Count);
+            }
+        }
+
+        [Fact(Skip = "RavenDB_8404")]
```
Comment (ayende) [inline test/FastTests/Client/Subscriptions/RavenDB-8404.cs:105]: why skipped?

**Review by ayende — CHANGES_REQUESTED** [review]

**Comment by maximburyak** [comment]
bahh.. the last commit was a mistake (forgot to switch to different branch)

File: src/Raven.Server/Documents/Queries/AST/JavascriptCodeQueryVisitor.cs (hunk 1 above)
Comment (maximburyak) [reply src/Raven.Server/Documents/Queries/AST/JavascriptCodeQueryVisitor.cs:102]: https://github.com/sebastienros/jint/issues/429

Comment (ayende) [reply src/Raven.Server/Documents/Queries/AST/JavascriptCodeQueryVisitor.cs:102]: Use polyfill:
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/startsWith

Comment (ayende) [reply src/Raven.Server/Documents/Queries/AST/JavascriptCodeQueryVisitor.cs:102]: However, regardless, we need to implement it directly.

File: src/Raven.Server/Documents/Queries/AST/JavascriptCodeQueryVisitor.cs (hunk 2 above)
Comment (maximburyak) [reply src/Raven.Server/Documents/Queries/AST/JavascriptCodeQueryVisitor.cs:114]: that's an idea

File: src/Raven.Server/Documents/Queries/AST/JavascriptCodeQueryVisitor.cs (hunk 3 above)
Comment (maximburyak) [reply src/Raven.Server/Documents/Queries/AST/JavascriptCodeQueryVisitor.cs:124]: actually, it works (except the <'> at the end, which caused an error). The query syntax also supports multiple parameters to "intersect"

File: src/Raven.Server/Documents/Queries/AST/JavascriptCodeQueryVisitor.cs (hunk 4 above)
Comment (maximburyak) [reply src/Raven.Server/Documents/Queries/AST/JavascriptCodeQueryVisitor.cs:143]: it's the same .net Regex object in both of them, the differences are really minor
https://github.com/sebastienros/jint/blob/dev/Jint/Native/RegExp/RegExpInstance.cs
https://github.com/sebastienros/jint/blob/dev/Jint/Native/RegExp/RegExpPrototype.cs#L80
https://github.com/ravendb/ravendb/blob/v4.0/src/Raven.Server/Documents/Queries/Regex/RegexTermEnum.cs

On the other hand, we would probably like to perform some regex object caching, so maybe we still have to implement it ourselves...

Comment (ayende) [reply src/Raven.Server/Documents/Queries/AST/JavascriptCodeQueryVisitor.cs:143]: We want the caching and we want to get the timeout behavior

Comment (maximburyak) [reply src/Raven.Server/Documents/Queries/AST/JavascriptCodeQueryVisitor.cs:143]: you mean cache timeout?

Comment (ayende) [reply src/Raven.Server/Documents/Queries/AST/JavascriptCodeQueryVisitor.cs:143]: MatchTimeout

**Pull request closed without merging**